Remove 'cluster_manager' role attachment when using 'node.master' deprecated setting

sandeshkr419 opened the pull request:
### Description
Using `node.master: true` configures both `cluster_manager` and `master` role to a node which is not correct, only one of them should be assigned.

Reference in cdk where it uses legacy settings to bootstrap nodes: https://github.com/opensearch-project/opensearch-cluster-cdk/blob/main/lib/opensearch-config/multi-node-base-config.yml

Allowing only one of the both roles to be attached instead of both - 'master' which is deprecated as part of change of 'master' nomenclature: https://github.com/opensearch-project/OpenSearch/issues/472 - since both the legacy setting `node.master` and `master` role are deprecated - so we need to keep the initial behavior only with deprecated roles and settings. 


### Issues Resolved
Resolves https://github.com/opensearch-project/OpenSearch/issues/6103

The change modifies the role attached in multi-node dedicated cluster-manager setup.

**API response before the change:**

Multi node with dedicated cluster-manager when bootstrapped with `node.master: true`:

```
..
"nodes" : {
    "count" : {
      "total" : 6,
      "cluster_manager" : 6,
      "coordinating_only" : 0,
      "data" : 3,
      "ingest" : 3,
      "master" : 6,
      "remote_cluster_client" : 6
    },
    "versions" : [
      "2.5.1"
    ],
..
```


```
curl "<cluster-endpoint>/_cat/nodes?v"
ip         heap.percent ram.percent cpu load_1m load_5m load_15m node.role node.roles                                   cluster_manager name
10.0.3.110           62          70   0    0.00    0.00     0.00 mmr       cluster_manager,master,remote_cluster_client -               manager-node
10.0.3.72            60          69   0    0.00    0.00     0.00 dir       data,ingest,remote_cluster_client            -               data-node
10.0.4.132           20          70   0    0.00    0.00     0.00 dir       data,ingest,remote_cluster_client            -               data-node
10.0.5.42            26          69   0    0.04    0.02     0.00 dir       data,ingest,remote_cluster_client            -               data-node
10.0.4.156           39          69   0    0.00    0.00     0.00 mmr       cluster_manager,master,remote_cluster_client -               manager-node
10.0.5.39            47          70   0    0.00    0.04     0.01 mmr       cluster_manager,master,remote_cluster_client *               seed
```

**API response after the change:**


Multi node with dedicated cluster-manager when bootstrapped with `node.master: true`:

```
..
"nodes" : {
    "count" : {
      "total" : 6,
      "cluster_manager" : 3,
      "coordinating_only" : 0,
      "data" : 3,
      "ingest" : 3,
      "master" : 3,
      "remote_cluster_client" : 6,
      "search" : 0
    },
..
```


```
curl   "<endpoint>/_cat/nodes?v"
ip         heap.percent ram.percent cpu load_1m load_5m load_15m node.role node.roles                            cluster_manager name
10.0.5.192           38          63   0    0.00    0.00     0.00 mr        master,remote_cluster_client -               manager-node
10.0.3.75            30          63   0    0.00    0.00     0.00 mr        master,remote_cluster_client -               manager-node
10.0.5.168           16          63   0    0.65    0.16     0.05 dir       data,ingest,remote_cluster_client     -               data-node
10.0.4.83            46          63   0    0.00    0.00     0.00 dir       data,ingest,remote_cluster_client     -               data-node
10.0.5.169           55          63   0    0.00    0.00     0.00 mr        master,remote_cluster_client *               seed
10.0.3.251           60          63   0    0.00    0.00     0.00 dir       data,ingest,remote_cluster_client     -               data-node
```

### Check List
- [ ] New functionality includes testing.
 - [ ] All tests pass
- [ ] New functionality has been documented.
- [ ] New functionality has javadoc added
- [ ] Commits are signed per the DCO using --signoff
- [ ] Commit changes are listed out in CHANGELOG.md file (See: [Changelog](../blob/main/CONTRIBUTING.md#changelog))

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

andrross commented:
At first glance, [adding the cluster manager role when node.master is true](https://github.com/opensearch-project/OpenSearch/blob/c1272c181e6bd96d1db725dfbbcd5ac8e6058502/server/src/main/java/org/opensearch/cluster/node/DiscoveryNodeRole.java#L248-L249) seems wrong. `node.master` is deprecated the same as adding `master` to `node.roles`, so it seems like that legacy setting should just add the `master` role and not the `cluster_manager` role. Can we just remove the linked code to fix this bug?

sandeshkr419 commented:
Thanks @andrross for providing feedback. I appreciate the ideation of keeping the deprecated legacy settings to be restricted to deprecated roles, in terms of selection of `master` and `cluster_manager` for the deprecated `node.master` setting. This makes sense with the understanding that in future OS 3.x release, `master` role and all legacy settings including `node.master` will be removed in entirely. 
I don't think removing this is within the scope of OS 2.x and specially in this PR. I tried removing this legacy setting from here and it breaks a whole lot of different things - which requires a complete code refactoring exercise which IMO is worth with OS 3.x where we deprecate this entirely. 

Here is the new response from `_cat/nodes` when the cluster is bootstrapped using legacy settings:
```
ip         heap.percent ram.percent cpu load_1m load_5m load_15m node.role node.roles                        cluster_manager name
10.0.5.117           21          61   0    0.00    0.01     0.03 dir       data,ingest,remote_cluster_client -               data-node
10.0.3.38            52          62   0    0.00    0.00     0.01 mr        master,remote_cluster_client      *               seed
10.0.4.97            26          61   0    0.00    0.00     0.00 mr        master,remote_cluster_client      -               manager-node
10.0.5.82            28          61   0    0.00    0.00     0.00 mr        master,remote_cluster_client      -               manager-node
10.0.4.191           27          62   0    0.00    0.00     0.00 dir       data,ingest,remote_cluster_client -               data-node
```
`_cluster/stats`:

```
..
...
 "nodes" : {
    "count" : {
      "total" : 5,
      "cluster_manager" : 3,
      "coordinating_only" : 0,
      "data" : 2,
      "ingest" : 2,
      "master" : 3,
      "remote_cluster_client" : 5,
      "search" : 0
    }
...
..
```
The tests added validate node roles and cluster stats response

I still want to point that the legacy settings: when you provide a combination of `node.master`, `node.data`, `node.ingest`, it sets up these 3 roles: `cluster_manager/master`, `data`, `ingest` **alongwith**  the 4th `remote_cluster_client` role for a node. `remote_cluster_client` doesn't gets setup automatically unless provided explicitly when using `node.roles` - I have tested this out by spinning up a cluster using cdk and providing these 3 node roles (data, ingest, master) instead of legacy settings - you get only the roles you specify and not the remote_cluster_client.

When setting up explicit roles - `master` or `cluster_manager` roles via `node.roles`, the behavior would continue to exist as before: the role will be provided as specified by the user.

shwetathareja commented on server/src/internalClusterTest/java/org/opensearch/action/admin/cluster/stats/ClusterStatsIT.java:
```diff
@@ -322,4 +323,81 @@ public void testFieldTypes() {
         }
     }
+
+    public void testNodeCountsWithLegacyMasterSetting() throws ExecutionException, InterruptedException {
```
please add test with cluster manager role as well

sandeshkr419 replied:
Thanks for checking. Added alongwith refactored some.

shwetathareja commented on server/src/internalClusterTest/java/org/opensearch/action/admin/cluster/stats/ClusterStatsIT.java:
```diff
@@ -322,4 +312,99 @@ public void testFieldTypes() {
+        ClusterStatsResponse clusterStatsResponse = client().admin().cluster().prepareClusterStats().get();
+        assertCounts(clusterStatsResponse.getNodesStats().getCounts(), total, expectedRoleCounts);
+
+        Set<String> expectedRoles = Set.of(DiscoveryNodeRole.MASTER_ROLE.roleName(),
```
why master role should be there in expectedRoles?

sandeshkr419 replied:
Thanks for pointing out. Missed one commit. Pushed with corrections.

andrross commented on server/src/internalClusterTest/java/org/opensearch/action/admin/cluster/stats/ClusterStatsIT.java:
```diff
@@ -322,4 +312,99 @@ public void testFieldTypes() {
+        assertEquals(expectedRoles, getNodeRoles(0));
+
+    }
+    private Map<String, Integer> getExpectedCounts(int dataRoleCount, int masterRoleCount, int clusterManagerRoleCount, int ingestRoleCount,
```
It looks like `searchRoleCount` and `coordinatingOnlyCount` are always zero, so you can omit them from the method signature and just set them to zero in the method body.

sandeshkr419 replied:
This is true for the existing tests. But keeping the method definition for future test cases instead of hardcoding the values. Adding those test-cases in this CR will be out of scope for these changes.

andrross commented on server/src/internalClusterTest/java/org/opensearch/action/admin/cluster/stats/ClusterStatsIT.java:
```diff
@@ -322,4 +312,99 @@ public void testFieldTypes() {
+
+    public void testNodeRolesWithClusterManagerRole() throws ExecutionException, InterruptedException {
+        int total = 1;
+        Settings legacyMasterSettings = Settings.builder()
```
`legacyMasterSettings` does not make sense as a variable name here

sandeshkr419 replied:
Ack. Renaming.

andrross reviewed (changes requested):
This PR is attempting to merge into the `2.5` branch. Please rebase it and open it against `main`. Once committed into `main` we will backport it into the necessary branches.

Also, I'm seeing failures like the following in the tests. Can you look in to that?

```
ClusterManagerNotDiscoveredException[null]
        at app//org.opensearch.action.support.clustermanager.TransportClusterManagerNodeAction$AsyncSingleAction$2.onTimeout(TransportClusterManagerNodeAction.java:348)
        at app//org.opensearch.cluster.ClusterStateObserver$ContextPreservingListener.onTimeout(ClusterStateObserver.java:394)
        at app//org.opensearch.cluster.ClusterStateObserver$ObserverClusterStateListener.onTimeout(ClusterStateObserver.java:294)
        at app//org.opensearch.cluster.service.ClusterApplierService$NotifyTimeout.run(ClusterApplierService.java:707)
        at app//org.opensearch.common.util.concurrent.ThreadContext$ContextPreservingRunnable.run(ThreadContext.java:747)
        at java.****@17.0.6/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
        at java.****@17.0.6/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
        at java.****@17.0.6/java.lang.Thread.run(Thread.java:833)
```

sandeshkr419 reviewed (commented):
Revising the code changes with new commit.

sandeshkr419 commented:
> This PR is attempting to merge into the `2.5` branch. Please rebase it and open it against `main`. Once committed into `main` we will backport it into the necessary branches.
> 
> Also, I'm seeing failures like the following in the tests. Can you look in to that?
> 
> ```
> ClusterManagerNotDiscoveredException[null]
>         at app//org.opensearch.action.support.clustermanager.TransportClusterManagerNodeAction$AsyncSingleAction$2.onTimeout(TransportClusterManagerNodeAction.java:348)
>         at app//org.opensearch.cluster.ClusterStateObserver$ContextPreservingListener.onTimeout(ClusterStateObserver.java:394)
>         at app//org.opensearch.cluster.ClusterStateObserver$ObserverClusterStateListener.onTimeout(ClusterStateObserver.java:294)
>         at app//org.opensearch.cluster.service.ClusterApplierService$NotifyTimeout.run(ClusterApplierService.java:707)
>         at app//org.opensearch.common.util.concurrent.ThreadContext$ContextPreservingRunnable.run(ThreadContext.java:747)
>         at java.****@17.0.6/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
>         at java.****@17.0.6/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
>         at java.****@17.0.6/java.lang.Thread.run(Thread.java:833)
> ```

@andrross Thanks for the comments and suggestions.
Rebased against latest `main` changes and fixed the broken test in latest commit.

sandeshkr419 commented:
@andrross Looking into other test failures, it seems like the logic of using `legacySettings` with cluster_manager as well is tightly coupled as of now as `cluster_uuid` is not getting assigned to the cluster when we remove [the setting of deprecated legacy settings](https://github.com/opensearch-project/OpenSearch/blob/c1272c181e6bd96d1db725dfbbcd5ac8e6058502/server/src/main/java/org/opensearch/cluster/node/DiscoveryNodeRole.java#L248-L249), which is causing multiple integration tests to fail as well.

```
 curl localhost:9200
{
  "name" : "runTask-0",
  "cluster_name" : "runTask",
  "cluster_uuid" : "_na_",
  "version" : {
    "distribution" : "opensearch",
    "number" : "3.0.0-SNAPSHOT",
    "build_type" : "tar",
    "build_hash" : "71893d15d1da33ca9fef19fb59d810448b6f231e",
    "build_date" : "2023-03-09T10:59:14.933571Z",
    "build_snapshot" : true,
    "lucene_version" : "9.5.0",
    "minimum_wire_compatibility_version" : "2.7.0",
    "minimum_index_compatibility_version" : "2.0.0"
  },
  "tagline" : "The OpenSearch Project: https://opensearch.org/"
}

```

Completely decoupling the deprecated `master` legacy setting from `cluster_manager` role seems outside the scope of this PR as this will require evaluating all the failed test cases along with adding new test cases as well. I'd suggest we remove the duplicate roles when [setting up roles from legacySettings](https://github.com/opensearch-project/OpenSearch/blob/c1272c181e6bd96d1db725dfbbcd5ac8e6058502/server/src/main/java/org/opensearch/cluster/node/DiscoveryNode.java#L287) in this PR and handle decoupling this in a separate PR.

andrross commented:
> Completely decoupling the deprecated `master` legacy setting from `cluster_manager` role seems outside the scope of this PR plus this will require evaluating all the failed test cases. I'd suggest we remove the duplicate roles when [setting up roles from legacySettings](https://github.com/opensearch-project/OpenSearch/blob/c1272c181e6bd96d1db725dfbbcd5ac8e6058502/server/src/main/java/org/opensearch/cluster/node/DiscoveryNode.java#L287) in this PR and handle decoupling this in 3.x / different PR.

I would love to get the whole issue fixed here. I'm not totally opposed to punting it if you can get the issue at hand fixed by another method.

However, I did spend a few minutes poking at this, and it appears the change in this PR broke the `DiscoveryNode::isEnabledByDefault` logic that uses legacy setting. I added the following to the `CLUSTER_MANAGER_ROLE` definition:

```
@Override
public boolean isEnabledByDefault(final Settings settings) {
    return settings.getAsBoolean("node.master", true);
}
```

That seemed to fix most of the tests that were relying on the "cluster manager by default" logic, but I ran into some failures. Maybe you could spend a few minutes looking at that to see if the bug is obvious?

sandeshkr419 commented:
@andrross 
> That seemed to fix most of the tests that were relying on the "cluster manager by default" logic, but I ran into some failures. Maybe you could spend a few minutes looking at that to see if the bug is obvious?

Still running locally `./gradlew run` doesn't identifies cluster_uuid / cluster_manager not found. + gets 102 tests failures as well.
I'm looking into the code flow to see what else can I find to fix this.

sandeshkr419 reviewed (commented):
Addressed comments by @andrross

sandeshkr419 commented:
@andrross The legacy master settings and cluster_manager role have been decoupled now to some extent. I have added relevant test cases to validate the same. I am not able to resolve on your "Changes Requested" - although I have rebased my changes against main.

Additionally, I have validated the changes manually as well by creating multi-node domains using node.roles and legacy settings both. 

@andrross @shwetathareja @tlfeng Requesting review for the same.

tlfeng reviewed (approved):
@sandeshkr419 Thanks for your effort in finding a properly solution to the problem. Seeing from the tests you added, the node role counts are correct - Assigning either `cluster_manager` or `master` role only results count 1 in node stats REST API.

andrross commented on server/src/main/java/org/opensearch/cluster/node/DiscoveryNodeRole.java:
```diff
@@ -273,6 +274,10 @@ public void validateRole(List<DiscoveryNodeRole> roles) {
 
+        @Override
+        public boolean isEnabledByDefault(final Settings settings) {
+            return !Booleans.isBoolean(settings.get("node.master"));
```
Nitpick, the the prevailing style is `Booleans.isBoolean(settings.get("node.master")) == false`

sandeshkr419 replied:
Thanks for reviewing @andrross! Addressed the comment and re-based my changes against main to resolve conflicts in *CHANGELOG.md*
